Sync/BatchingClient.swift

```diff
@@ -0,0 +1,263 @@
+         * When talking about a 'batch' and multiple 'batches' the word becomes very confusing. In this context,
+         *
+         * A 'batch' consists of N requests to the server beginning with a batch=true call and ends with a commit=true call.
+         * A 'bunch' is many of these calls.
+         * A partialBatch is one of the N requests made within a 'batch'.
+         *
+         * Here are some typealiases to make your life easier.
+         */
+
+        typealias Batch = [[String]]
+        typealias Bunch = [Batch]
+
+        var bunch = Bunch()
+
+        while !mutBatch.isEmpty {
+            var recordCount = 0
+            var size = 0
+            var smallerBatch = Batch()
+
+            while size <= config.maxTotalBytes && recordCount <= config.maxTotalRecords && !mutBatch.isEmpty {
+                let partialBatch = mutBatch.removeFirst()
+                recordCount += partialBatch.count
+                size += partialBatch.reduce(0) { $0 + ($1.utf8.count + 1) }
+                smallerBatch.append(partialBatch)
+            }
+
+            bunch.append(smallerBatch)
+        }
+
+        let perBatch: (batch: [[String]]) -> Success = { batch in
+            return self.batchUpload(batch, ifUnmodifiedSince: ifUnmodifiedSince, onCollectionUploaded: onCollectionUploaded)
+        }
+
+        return walk(bunch, f: perBatch)
+    }
+
+    public func batchUpload(batch: [[String]], ifUnmodifiedSince: Timestamp?, onCollectionUploaded: (POSTResult -> Void)) -> Success {
+
+        let batchSize: ByteCount = batch.reduce(0) { sum, batch in
+            return sum + batch.reduce(0) { $0 + $1.utf8.count }
+        }
+
+        let recordCount: Int = batch.reduce(0) { $0 + $1.count }
+
+        // Same checks made in commit call but included here for strength
+        precondition(recordCount < config.maxTotalRecords)
+        precondition(batchSize < config.maxTotalBytes)
+
+        var batch = batch
+        let firstBatch = batch.removeFirst()
+        let lastBatch = batch.last ?? []
+
+        return startBatch(firstBatch).bind { result in
+            guard let token = result.successValue else {
+                // Check if we didn't get a token back/batching isn't supported and push up records using regular single posts
+                if let error = result.failureValue as? BatchingNotSupported {
+
+                    // Since we uploaded the first set already as part of the start call and we don't support
+                    // batching, make sure to invoke the upload callback to let the others know we committed
+                    // new records to the collection server-side
+                    onCollectionUploaded(error.response)
+
+                    // Walk through the batches, posting along the way and invoking onUpload
+                    let perChunk: (lines: [String]) -> Success = { lines in
+                        return self.uploader(lines: lines, ifUnmodifiedSince: ifUnmodifiedSince, queryParams: nil)
```

Comment:
I think this `ifUnmodifiedSince` is wrong. You should use the timestamp from the batch that we just uploaded.